cow_instrument/DetectorInfo.h

```diff
@@ -232,24 +281,98 @@ template <typename InstTree> size_t DetectorInfo<InstTree>::size() const {
 }
 
 template <typename InstTree>
+size_t DetectorInfo<InstTree>::componentSize() const {
+  return m_instrumentTree->componentSize();
+}
+
+template <typename InstTree>
 const InstTree &DetectorInfo<InstTree>::const_instrumentTree() const {
   return *m_instrumentTree;
 }
 
 template <typename InstTree>
-void DetectorInfo<InstTree>::modify(size_t nodeIndex, Command &command) {
+void DetectorInfo<InstTree>::moveComponent(size_t componentIndex,
+                                           const Eigen::Vector3d &offset) {
+
+  const std::vector<size_t> indexes =
+      m_instrumentTree->subTreeIndexes(componentIndex);
+  for (auto &index : indexes) {
+    (*m_positions)[index] += offset;
+    // TODO m_entry and m_exit points should also be translated here!
+  }
+}
+
+template <typename InstTree>
+void DetectorInfo<InstTree>::move(size_t componentIndex,
+                                  const Eigen::Vector3d &offset) {
 
-  m_instrumentTree->modify(nodeIndex, command);
+  moveComponent(componentIndex, offset);
 
   // All other geometry-derived information is now also invalid. Very
   // important!
-  initL1();
-  initL2();
+  init();
+
+  // Meta-data should all still be valid.
+}
+
+template <typename InstTree>
+void DetectorInfo<InstTree>::rotate(size_t componentIndex,
+                                    const Eigen::Vector3d &axis,
+                                    const double &theta,
+                                    const Eigen::Vector3d &center) {
+
+  rotateComponent(componentIndex, axis, theta, center);
+  // All other geometry-derived information is now also invalid. Very
+  // important!
+  init();
 
   // Meta-data should all still be valid.
 }
 
 template <typename InstTree>
+void DetectorInfo<InstTree>::rotateComponent(size_t componentIndex,
+                                             const Eigen::Vector3d &axis,
+                                             const double &theta,
+                                             const Eigen::Vector3d &center) {
+
+  using namespace Eigen;
+  const auto transform =
+      Translation3d(center) * AngleAxisd(theta, axis) * Translation3d(-center);
+  const auto rotation = transform.rotation();
+
+  const std::vector<size_t> indexes =
+      m_instrumentTree->subTreeIndexes(componentIndex);
+  for (auto &index : indexes) {
+    (*m_positions)[index] = transform * (*m_positions)[index];
+    (*m_rotations)[index] = rotation * (*m_rotations)[index];
+    // TODO m_entry and m_exit points should also be translated here!
+  }
+}
+
+template <typename InstTree>
+void DetectorInfo<InstTree>::moveDetector(size_t detectorIndex,
+                                          const Eigen::Vector3d &offset) {
+
+  moveComponent(m_instrumentTree->detIndexToCompIndex(detectorIndex), offset);
+
+  // Only l2 needs to be recalculated.
+  initL2();
```

Comment:
Only a single detector is moved, so there is no need to rebuild the full L2 vector, I think?